src/aks-preview/azext_aks_preview/_roleassignments.py

```diff
@@ -0,0 +1,136 @@
+    else:
+        scope = subscription_scope
+    return scope
+
+
+def create_role_assignment(cli_ctx, role, assignee,
+                           is_service_principal=True, resource_group_name=None, scope=None, resolve_assignee=True):
+    return _create_role_assignment(cli_ctx,
+                                   role, assignee, resource_group_name,
+                                   scope, resolve_assignee=(is_service_principal and resolve_assignee))
+
+
+def _create_role_assignment(cli_ctx, role, assignee,
+                            resource_group_name=None, scope=None, resolve_assignee=True):
+    from azure.cli.core.profiles import ResourceType, get_sdk
+    factory = get_auth_management_client(cli_ctx, scope)
+    assignments_client = factory.role_assignments
+    definitions_client = factory.role_definitions
+
+    scope = build_role_scope(
+        resource_group_name, scope, assignments_client.config.subscription_id)
+
+    # XXX: if role is uuid, this function's output cannot be used as role assignment defintion id
+    # ref: https://github.com/Azure/azure-cli/issues/2458
+    role_id = resolve_role_id(role, scope, definitions_client)
+
+    # If the cluster has service principal resolve the service principal client id to get the object id,
+    # if not use MSI object id.
+    object_id = resolve_object_id(
+        cli_ctx, assignee) if resolve_assignee else assignee
+    RoleAssignmentCreateParameters = get_sdk(cli_ctx, ResourceType.MGMT_AUTHORIZATION,
+                                             'RoleAssignmentCreateParameters', mod='models',
+                                             operation_group='role_assignments')
+    parameters = RoleAssignmentCreateParameters(
+        role_definition_id=role_id, principal_id=object_id)
+    assignment_name = uuid.uuid4()
+    custom_headers = None
+    return assignments_client.create(scope, assignment_name, parameters, custom_headers=custom_headers)
+
+
+def add_role_assignment(cli_ctx, role, service_principal_msi_id, is_service_principal=True, delay=2, scope=None):
+    # AAD can have delays in propagating data, so sleep and retry
+    hook = cli_ctx.get_progress_controller(True)
+    hook.add(message='Waiting for AAD role to propagate',
+             value=0, total_val=1.0)
+    logger.info('Waiting for AAD role to propagate')
+    for x in range(0, 10):
+        hook.add(message='Waiting for AAD role to propagate',
+                 value=0.1 * x, total_val=1.0)
+        try:
+            # TODO: break this out into a shared utility library
+            create_role_assignment(
+                cli_ctx, role, service_principal_msi_id, is_service_principal, scope=scope)
+            break
+        except CloudError as ex:
+            if ex.message == 'The role assignment already exists.':
+                break
+            logger.info(ex.message)
+        except:  # pylint: disable=bare-except
+            pass

```

Comment:
at least one logging.

Reply:
Hi Andy, a lot of this code is not mine. It was simply moved around for organization purposes and to avoid circular dependencies. I ensured everything still works, but did not change any code functionality except for functions that were directly related to addons (and there were only 2 of them). I'm not sure it would be safe for me to change the functionality of these other functions without knowing there purpose.